bam_to_mate_hist.py

```diff
@@ -11,147 +11,291 @@
+            continue
+
+        last_read = read.qname
+        ref1 = read.reference_id
+        ref2 = read.next_reference_id
+
+        if ref1 != ref2:
+            diff_chr += 1
+            if count_diff_refname_stub:
+                ref1_stub = refs[ref1].split(".")[0]
+                ref2_stub = refs[ref2].split(".")[0]
+                # print ref1_stub, ref2_stub
+                if ref1_stub != ref2_stub:
+                    diff_stub += 1
+
+        else:
+            read1_pos = read.reference_start
+            read2_pos = read.next_reference_start
+            dist = str(abs(read1_pos - read2_pos))
+            dists[num] = dist
+        num += 1
+
+    dists = dists[0:num + 1]  # why am i doing this?
+    return diff_chr, dists, diff_stub, split_reads, dupe_reads
+
+
+def calc_n50_from_header(header, xx=50.0):
+    '''calculate the N50 of the starting assembly from the information in a pysam header object.
+
+        Args:
+            header (pysam.AlignmentFile.header): the header from which to extract reference sequence lengths
+            xx (float): the XX of NXX. we assume it is N50 so the default is 50.0
+
+        Returns:
+            contig_len (int): the NXX (probably N50) of the assembly based on the header
+            total (int): the total length of the assembly
+
+    '''
+    frac = xx / 100.0
+    lens = [contig["LN"] for contig in header["SQ"]]
+    lens.sort()
+    total = sum(lens)
+    nxx_len = total * frac
+
+    cumsum = 0
+    contig_len = 0
+    for length in reversed(lens):
+        contig_len = length
+        cumsum += length
+        if cumsum >= nxx_len:
+            break
+    return contig_len, total
+
+
+def make_histograms(dists, bamfile_handle, num_reads, outfile_name):
+    '''make the read distance histograms using matplotlib and write them to disk.
 	Args:
 		dists (numpy array of ints): Distances to plot in histogram.
 		bamfile_handle (str): path to bamfile of dists
-		'''
-	num_reads = len(dists)
-	with PdfPages(outpdf_name) as pdf:
-		fig1 = plt.figure()
-		plt.hist(dists, bins=40)
-		ax = fig1.add_subplot(111)
-		ax.set_ylim(0.5, num_reads * 2)
-		plt.yscale("log", nonposy="clip")
-		plt.title("Mate distance distribution for sample\n" + bamfile_handle+"\nfor first " + str(num_reads)+ " reads")
-		plt.xlabel("Distance between read pair mates in Hi-C mapping")
-		plt.ylabel("Number of reads")
-		pdf.savefig(fig1)
-		plt.close()
-
-		fig2 = plt.figure()
-		plt.hist(dists, bins=xrange(0,20000, 500))
-		ax = fig2.add_subplot(111)
-		ax.set_xlim(0, 20000)
-		ax.set_ylim(0.5, num_reads * 2)
-		plt.yscale("log", nonposy="clip")
-		plt.title("Mate distance distribution for sample\n" + bamfile_handle+"\nfor first " + str(num_reads)+ " reads")
-		plt.xlabel("Distance between read pair mates in Hi-C mapping")
-		plt.ylabel("Number of reads")
-		pdf.savefig(fig2)
-		plt.close()
+	'''
+    num_reads = len(dists)
+    # with PdfPages(outfile_name) as pdf:
+    fig1 = plt.figure()
```

Comment:
i have no attachments. this is the first (useful) thing i've ever done with matplotlib. i feel like matplotlib support from numpy would be better, but i have no data to back that up. i am equally ignorant of plotly.

Reply:
Zev and I like to use plotly.  We can look into the numpy support, but I suspect it isn't hard and Zev can likely tell us directly without looking.  The biggest problem with matplotlib can be the look of the plots and you got some nice ones, so that isn't a concern here.  For the web, we'd want an interactive web library, like plotly, but that isn't required for the current application.

Reply:
We should as a company decide on a default plotting library we try to prototype in with Python, if possible.  I know some libraries are better or worse than others for different types of plots, but where feasible, we should try to use the same library.  The fact plots may be targeted for the web makes me prefer plotly to save on double effort.  I also like the API for plotly better than matplotlib.